File: binary_tree_DFS/path_sum_II.py
#  declear a list 
#  base case: if node == Null --> return nothing
#  if node == leaf node and sum == targetSum --> return list
#  if node is not a leaf node, add the value into sum,  add value to 

def pathSum(root, targetSum):
    result = []
    def getList(node, currSum, resultList):
        if not node:
            return
        else:
            currSum += node.val
            resultList.append(node.val)
            if not node.left and not node.right:
                if currSum == targetSum:
                    result.append(list(resultList))
                       
            getList(node.left, currSum, resultList) 
            getList(node.right, currSum, resultList)

            resultList.pop()
                
    getList(root, 0, [])
    return result

File: binary_tree_DFS/path_sum_II.js
const pathSum = (root, targetSum) => {
    const result = []
    const dfs = (node, currSum, resultArr) => {
        if (node === null) return;
        currSum += node.val;
        resultArr.push(node.val);
        if (node.left === null && node.right === null) {
            if (currSum === targetSum)
            result.push(resultArr.slice())
        }

        dfs(node.left, currSum, resultArr)
        dfs(node.right, currSum, resultArr)

        resultArr.pop()
    }

    dfs(root, 0, [])
    return result;
}